{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0b7de815",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter x : 10\n",
      "Enter y : 20\n",
      "Result :  30\n",
      "Code inside try block is executed successfully\n"
     ]
    }
   ],
   "source": [
    "1. What is the role of the 'else' block in a try-except statement? Provide an example scenario where it would be useful.\n",
    "- Else block in exception handling is executed when there is no exception raised within the try block. \n",
    "- The purpose of this block is to define the code inside the try block is executed successfully without any error.\n",
    "Example:\n",
    "    \n",
    "def addition():\n",
    "    try:\n",
    "        x = int(input(\"Enter x : \"))\n",
    "        y = int(input(\"Enter y : \"))\n",
    "        z = x + y\n",
    "        print(\"Result : \", z)\n",
    "    except ValueError:\n",
    "        print(\"Invalid input. Please enter integers only.\")\n",
    "    else:\n",
    "        print(\"Code inside try block is executed successfully\")\n",
    "addition()       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a6daf5ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the numerator: 10\n",
      "Enter the denominator: 0\n",
      "Cannot divide by zero.\n",
      "Inner try block executed.\n",
      "Outer try block executed.\n"
     ]
    }
   ],
   "source": [
    "2. Can a try-except block be nested inside another try-except block? Explain with an example.\n",
    "- Yes, a try-except block can be nested inside another try-except block this is called nested exception handling.\n",
    "- It allows for more fine-grained handling of exceptions and the ability to handle exceptions at different levels of code.\n",
    "Example:\n",
    "    \n",
    "def nested_exception() :   \n",
    "    try:# Outer try block\n",
    "        num1 = int(input(\"Enter the numerator: \"))  \n",
    "        try:  # Inner try block\n",
    "            num2 = int(input(\"Enter the denominator: \"))\n",
    "            result = num1 / num2\n",
    "            print(\"Result:\", result)\n",
    "        except ValueError:\n",
    "            print(\"Value Error invalid denominator. Please enter a number.\")\n",
    "        except ZeroDivisionError:\n",
    "            print(\"Cannot divide by zero.\")\n",
    "        finally:\n",
    "            print(\"Inner try block executed.\")\n",
    "    except ValueError:\n",
    "        print(\"Value Error invalid denominator. Please enter a number.\")\n",
    "    finally:\n",
    "        print(\"Outer try block executed.\")\n",
    "nested_exception()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d78d13d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a string: Apple\n",
      "Error: 'Apple'is not a palindrome\n"
     ]
    }
   ],
   "source": [
    "3. How can you create a custom exception class in Python? Provide an example that demonstrates its usage.\n",
    "A custom exception class can be created by defining a new class that inherits from the built-in Exception class or any of its subclasses.\n",
    "Example:\n",
    "\n",
    "class NotPalindromeException(Exception):\n",
    "    pass\n",
    "\n",
    "def check_palindrome(string):\n",
    "    if string.lower() != string.lower()[::-1]:\n",
    "        raise NotPalindromeException(f\"'{string}'is not a palindrome\")\n",
    "\n",
    "try:\n",
    "    input_string = input(\"Enter a string: \")\n",
    "    check_palindrome(input_string)\n",
    "    print(f\"'{input_string}' is a palindrome.\")\n",
    "except NotPalindromeException as e:\n",
    "    print(\"Error:\", str(e))\n",
    "except Exception as e:\n",
    "    print(\"An unexpected error occurred:\", str(e))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a5f5e0f4",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid decimal literal (876181773.py, line 132)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  Cell \u001b[1;32mIn[3], line 132\u001b[1;36m\u001b[0m\n\u001b[1;33m    14.IndentationError: Raised when there is an incorrect indentation in the code.\u001b[0m\n\u001b[1;37m      ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid decimal literal\n"
     ]
    }
   ],
   "source": [
    "4. What are some common exceptions that are built-in to Python?\n",
    "\n",
    "Python provides several built-in exceptions that cover a wide range of error scenarios. Some common built-in exceptions in Python include:\n",
    "\n",
    "1. TypeError : \n",
    "    Raised when an operation or function is performed on an object of an inappropriate type.\n",
    "Example:\n",
    "try:\n",
    "    x = \"10\"  # Assigning a string value\n",
    "    y = 2\n",
    "    z = x + y  # addition between a string and an integer\n",
    "    print(\"Result:\", z)\n",
    "except TypeError:\n",
    "    print(\"Error: TypeError occurred.\")\n",
    "\n",
    "\n",
    "2. ValueError : \n",
    "    Raised when a function receives an argument of the correct type but an invalid value.\n",
    "Example:\n",
    "try:\n",
    "    n = int(input(\"Please enter the numerator: \"))\n",
    "    d = int(input(\"Please enter the denominator: \"))\n",
    "    result = n / d\n",
    "    print(\"Result:\", result)\n",
    "except ValueError:\n",
    "    print(\"Please enter valid integers for the numerator and denominator.\")    \n",
    "\n",
    "\n",
    "3. NameError: \n",
    "    Raised when a local or global name is not found.\n",
    "Example:\n",
    "try:\n",
    "    side = \"4\"\n",
    "    print(area)  # Attempting to access an undefined variable\n",
    "except NameError:\n",
    "    print(\"NameError is there\")\n",
    "    \n",
    "    \n",
    "4. IndexError: Raised when a sequence subscript is out of range.\n",
    "Example:\n",
    "try:\n",
    "    list1 = [1, 2, 3]\n",
    "    print(list1[3])  # Trying to reach an index\n",
    "except IndexError:\n",
    "    print(\"IndexError error occurred.\")\n",
    "\n",
    "    \n",
    "5. KeyError: Raised when a dictionary key is not found.\n",
    "Example:\n",
    "try:\n",
    "    my_dictionary = {\"name\": \"John\", \"age\": 30}\n",
    "    result\n",
    "    print(my_dictionary[\"city\"])  # Accessing a KEY that does not exist\n",
    "except KeyError:\n",
    "    print(\"KeyError error occurred.\")\n",
    "\n",
    "    \n",
    "6. FileNotFoundError: Raised when a file or directory is requested, but it cannot be found.\n",
    "Example:\n",
    "try:\n",
    "    file_location = \"file.txt\"\n",
    "    with open(file_path, \"r\") as file:\n",
    "        contents = file.read()\n",
    "except FileNotFoundError:\n",
    "    print(f\"File '{file_path}' not found.\")\n",
    "\n",
    "    \n",
    "7. ZeroDivisionError: Raised when division or modulo operation is performed with a zero divisor.\n",
    "Example:\n",
    "n = int(input(\"Please enter the numerator: \"))\n",
    "d = int(input(\"Please enter the denominator: \"))\n",
    "result = n / d\n",
    "print(\"Result:\", result)\n",
    "\n",
    "\n",
    "8. IOError : Raised when an input/output operation fails.\n",
    "Example:\n",
    "try:\n",
    "    file_location = \"file.txt\"\n",
    "    with open(file_location, \"w\") as file:\n",
    "        file.write(\"This is my file\")\n",
    "except IOError:\n",
    "    print(f\"Unable to write to file '{file_location}'\")\n",
    "\n",
    "    \n",
    "9. AttributeError: Raised when an attribute reference or assignment fails.\n",
    "Example:\n",
    "try:\n",
    "    age = 20\n",
    "    # Using append method\n",
    "    age.append(12)\n",
    "except OverflowError:\n",
    "    print(\"'age' object has no attribute\")\n",
    "else: \n",
    "    print(result)\n",
    "\n",
    "    \n",
    "10. ImportError: Raised when an import statement fails to find and load a module.\n",
    "Example:\n",
    "try:\n",
    "    import Error\n",
    "except ImportError:\n",
    "    print(\" Import Error\")\n",
    "\n",
    "    \n",
    "11. OverflowError: Raised when the result of an arithmetic operation is too large to be represented.\n",
    "Example:\n",
    "import math\n",
    "try:\n",
    "    factorial = math.factorial(1000)\n",
    "    print(factorial)\n",
    "except OverflowError as error:\n",
    "    print(\"Error:\", error)\n",
    "\n",
    "\n",
    "12. MemoryError: Raised when an operation cannot be completed due to insufficient memory.\n",
    "Example:\n",
    "try:\n",
    "    large_list = [0] * (10**9)  # Creating a list with one billion elements\n",
    "    print(\"List created successfully.\")\n",
    "except MemoryError as error:\n",
    "    print(\"Error:\", error)\n",
    "    \n",
    "    \n",
    "13. SyntaxError: Raised when there is a syntax error in the code.\n",
    "Example:\n",
    "try:\n",
    "    print \"Hello, World!\"\n",
    "except SyntaxError:\n",
    "    print(\"Error: Invalid syntax\")\n",
    "    \n",
    "    \n",
    "14.IndentationError: Raised when there is an incorrect indentation in the code.\n",
    "Example:\n",
    "try:\n",
    "name = \"John\"\n",
    "print(name)\n",
    "except IndentationError:\n",
    "    print(\"IndentationError\")\n",
    "\n",
    "    \n",
    "15. KeyboardInterrupt : Raised when the user interrupts the execution of the program by pressing Ctrl+C.\n",
    "Example:\n",
    "try:\n",
    "    while True:\n",
    "        # Perform some repetitive task\n",
    "        print(\"Task in progress...\")\n",
    "except KeyboardInterrupt:\n",
    "    print(\"Task interrupted by the user.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f159aa28",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-06-27 10:44:08,515 - DEBUG - Varible are 10 and 15\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "5. What is logging in Python, and why is it important in software development?\n",
    "-Logging is the process of storing informative messages or events during the execution of a program.\n",
    "-It stores the programs state, actions, and potential errors in a log file.\n",
    "\n",
    "Logging is important in software development for several reasons:\n",
    "\n",
    "Debugging and troubleshooting: \n",
    "    Logging allows developers to trace the flow of execution, identify issues, and debug problems more effectively.\n",
    "\n",
    "Error and exception tracking:\n",
    "    It allows developers to capture valuable information, such as error messages, stack traces, and context, which can be used for diagnosing and resolving issues.\n",
    "\n",
    "Performance monitoring: \n",
    "    Logging can include performance-related metrics and measurements, such as execution times, resource usage, and throughput.     \n",
    "    \n",
    "Long-term analysis and maintenance:\n",
    "    Logging provides historical records that can be used for long-term analysis, identifying patterns, trends, and recurring issues. \n",
    "    \n",
    "    \n",
    "# Example:\n",
    "import logging\n",
    "logging.basicConfig(level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')\n",
    "\n",
    "def add(x,y):\n",
    "    logging.debug(\"Varible are %s and %s\", x,y)\n",
    "    return x+y\n",
    "add(10,15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fb5b9773",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:root:Starting the program\n",
      "DEBUG:root:Calculating sum of 10 and 5\n",
      "DEBUG:root:Sum: 15\n",
      "DEBUG:root:Total: 15\n",
      "DEBUG:root:Program finished\n",
      "INFO:root:Application starting up...\n",
      "INFO:root:Application shutting down...\n",
      "WARNING:root:Invalid input. Expected a list.\n",
      "ERROR:root:File not found: path/to/missing/file.txt\n",
      "CRITICAL:root:Critical error occurred. System shutting down.\n"
     ]
    }
   ],
   "source": [
    "6. Explain the purpose of log levels in Python logging and provide examples of when each log level would be appropriate.\n",
    "- Log levels in Python logging provide a way to categorize log messages based on their importance or severity. \n",
    "- Here are the commonly used log levels\n",
    "\n",
    "DEBUG:\n",
    "    It is typically used during development and debugging to capture fine-grained details that help identify issues or trace program execution.\n",
    "Example:\n",
    "import logging\n",
    "\n",
    "logging.basicConfig(level=logging.DEBUG)\n",
    "logger = logging.getLogger()\n",
    "\n",
    "def calculate_sum(a, b):\n",
    "    logger.debug(f\"Calculating sum of {a} and {b}\")\n",
    "    result = a + b\n",
    "    logger.debug(f\"Sum: {result}\")\n",
    "    return result\n",
    "\n",
    "def main():\n",
    "    logger.debug(\"Starting the program\")\n",
    "    num1 = 10\n",
    "    num2 = 5\n",
    "    total = calculate_sum(num1, num2)\n",
    "    logger.debug(f\"Total: {total}\")\n",
    "    logger.debug(\"Program finished\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n",
    "\n",
    "INFO: \n",
    "    -The INFO level is used for general information messages that highlight the progress of the program or significant events. \n",
    "    -These messages are typically used to provide high-level insights into the program's behavior.  \n",
    "Example:\n",
    "import logging\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "logger = logging.getLogger()\n",
    "\n",
    "def startup():\n",
    "    logger.info(\"Application starting up...\")\n",
    "    # Perform necessary initialization tasks\n",
    "\n",
    "def shutdown():\n",
    "    # Perform cleanup tasks\n",
    "    logger.info(\"Application shutting down...\")\n",
    "\n",
    "def main():\n",
    "    startup()\n",
    "    # Application logic\n",
    "    shutdown()\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n",
    "\n",
    "    \n",
    "WARNING: \n",
    "    -The WARNING level is used for non-critical issues or potential problems that may need attention. \n",
    "    -These messages indicate that something unexpected occurred but the program can continue running. \n",
    "Example:\n",
    "import logging\n",
    "\n",
    "logging.basicConfig(level=logging.WARNING)\n",
    "logger = logging.getLogger()\n",
    "\n",
    "def process_data(data):\n",
    "    if not isinstance(data, list):\n",
    "        logger.warning(\"Invalid input. Expected a list.\")\n",
    "        # Handle the issue gracefully or raise an exception\n",
    "        return\n",
    "\n",
    "    # Perform data processing steps\n",
    "\n",
    "def main():\n",
    "    data = \"abc\"\n",
    "    process_data(data)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n",
    "\n",
    "    \n",
    "ERROR: \n",
    "    -The ERROR level is used to log error conditions that may prevent the program from functioning correctly. \n",
    "    -These messages indicate more severe issues but do not necessarily halt the program.   \n",
    "Example:\n",
    "import logging\n",
    "\n",
    "logging.basicConfig(level=logging.ERROR)\n",
    "logger = logging.getLogger()\n",
    "\n",
    "def read_file(file_path):\n",
    "    try:\n",
    "        with open(file_path, 'r') as file:\n",
    "            content = file.read()\n",
    "            # Process the file content\n",
    "    except FileNotFoundError:\n",
    "        logger.error(f\"File not found: {file_path}\")\n",
    "\n",
    "def main():\n",
    "    file_path = \"path/to/missing/file.txt\"\n",
    "    read_file(file_path)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n",
    "\n",
    "    \n",
    "CRITICAL:\n",
    "    -The CRITICAL logging level represents the highest level of severity. \n",
    "    -It is used to indicate critical errors that may lead to the termination of the program.\n",
    "Example \n",
    "import logging\n",
    "\n",
    "logging.basicConfig(level=logging.CRITICAL)\n",
    "logger = logging.getLogger()\n",
    "\n",
    "def perform_critical_operation(critical_condition):\n",
    "    # Perform critical operation\n",
    "    if critical_condition:\n",
    "        logger.critical(\"Critical error occurred. System shutting down.\")\n",
    "        # Terminate the program or take appropriate actions\n",
    "\n",
    "def main():\n",
    "    perform_critical_operation(True)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7f793408",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:This is an informational message\n"
     ]
    }
   ],
   "source": [
    "7. What are log formatters in Python logging, and how can you customise the log message format using formatters?\n",
    "-Log formatters are used to define the structure and content of log messages. \n",
    "-This format, which shows the level, name, and message separated by a colon (:).\n",
    "-The default output format that can be configured to include things like timestamp, line number, and other details.\n",
    "\n",
    "Example\n",
    "\n",
    "import logging\n",
    "\n",
    "# Configure the logging settings\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "# Create a custom log formatter\n",
    "formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')\n",
    "\n",
    "# Create a file handler and set the formatter\n",
    "file_handler = logging.FileHandler('error1.log')\n",
    "file_handler.setFormatter(formatter)\n",
    "\n",
    "\n",
    "logger = logging.getLogger()\n",
    "logger.addHandler(file_handler)\n",
    "\n",
    "\n",
    "logger.info('This is an informational message')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01eabb0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "8. How can you set up logging to capture log messages from multiple modules or classes in a Python application?\n",
    "-To capture log messages from multiple modules or classes in a Python application, you can set up a logger hierarchy using the logging module. \n",
    "-This allows you to configure different loggers for different parts of your application and control the logging behavior at each level. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cda1b8f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# module1.py\n",
    "import logging\n",
    "\n",
    "# Create a logger for module1\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "def do_something():\n",
    "    # Log a message from module1\n",
    "    logger.info('This is a message from module1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4a98b06e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# module2.py\n",
    "import logging\n",
    "\n",
    "# Create a logger for module2\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "def do_something_else():\n",
    "    # Log a message from module2\n",
    "    logger.info('This is a message from module2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e5b3b9be",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'module1'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[15], line 4\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# Example\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;66;03m# main.py\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mlogging\u001b[39;00m\n\u001b[1;32m----> 4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mmodule1\u001b[39;00m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mmodule2\u001b[39;00m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;66;03m# Configure the logging settings\u001b[39;00m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'module1'"
     ]
    }
   ],
   "source": [
    "# Example\n",
    "# main.py\n",
    "import logging\n",
    "import module1\n",
    "import module2\n",
    "\n",
    "# Configure the logging settings\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "# Create a logger for the main module\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "# Log a message from the main module\n",
    "logger.info('This is a message from the main module')\n",
    "\n",
    "# Call functions or methods from other modules\n",
    "module1.do_something()\n",
    "module2.do_something_else()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1ffae0b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-06-26 10:33:25,699 - CRITICAL - Memory allocation failed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The value of x is: 10\n"
     ]
    }
   ],
   "source": [
    "9. What is the difference between the logging and print statements in Python? When should you use logging over print statements in a real-world application?\n",
    "Logging:\n",
    "    -Logging allows to capture important events, errors, warnings, or informational messages in a structured manner. \n",
    "    -Logging provides features like log levels, log formatting, and the ability to log to various outputs such as files, the console, or remote servers.\n",
    "    -It is designed for use in production environments and long-term maintenance of applications.\n",
    "Example:\n",
    "import logging\n",
    "logging.basicConfig(level=logging.CRITICAL)\n",
    "logging.critical('Memory allocation failed') \n",
    "\n",
    "\n",
    "Print Statement:\n",
    "    -The print statement is a simple way to display information on the console during program execution. \n",
    "    -It is often used for quick debugging, inspecting variable values, or providing immediate feedback.\n",
    "Example:\n",
    "x = 10\n",
    "print('The value of x is:', x)\n",
    "\n",
    "We use logging when, we can customize what information is to be printed along with the actual message. \n",
    "The information that we can print includes the package name, log level, line number, timestamp, method name, etc.\n",
    "\n",
    "With print it would require a lot of manual work to customize information, \n",
    "while logging frameworks provide loggers, to simply define a pattern in the logger configuration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c5020c8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Hello, World!\n"
     ]
    }
   ],
   "source": [
    "# 10. Write a Python program that logs a message to a file named \"app.log\" with the following requirements: \n",
    "#     ● The log message should be \"Hello, World!\"\n",
    "#     ● The log level should be set to \"INFO.\" \n",
    "#     ● The log file should append new log entries without overwriting previous ones.\n",
    "# Example:\n",
    "import logging\n",
    "\n",
    "# Configure the logging settings\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "# Create a custom log formatter\n",
    "formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')\n",
    "# Create a file handler and set the formatter\n",
    "file_handler = logging.FileHandler('app.log')\n",
    "file_handler.setFormatter(formatter)\n",
    "\n",
    "\n",
    "logger = logging.getLogger()\n",
    "logger.addHandler(file_handler)\n",
    "\n",
    "def print_info():\n",
    "    logger.info('Hello, World!')\n",
    "print_info()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8a17b5ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:__main__:Exception: ZeroDivisionError - Timestamp: 2023-06-29 10:30:58.144405\n"
     ]
    }
   ],
   "source": [
    "# 11. Create a Python program that logs an error message to the console and a file named \"errors.log\" \n",
    "# if an exception occurs during the program's execution. The error message should include the exception type and a timestamp.\n",
    "import logging\n",
    "import datetime\n",
    "\n",
    "# Configure the logging settings\n",
    "logging.basicConfig(level=logging.ERROR)\n",
    "\n",
    "# Create a file handler\n",
    "file_handler = logging.FileHandler('errors.log')\n",
    "file_handler.setLevel(logging.ERROR)\n",
    "\n",
    "# Create a console handler\n",
    "console_handler = logging.StreamHandler()\n",
    "console_handler.setLevel(logging.ERROR)\n",
    "\n",
    "# Create a formatter\n",
    "formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')\n",
    "\n",
    "# Set the formatter for the handlers\n",
    "file_handler.setFormatter(formatter)\n",
    "console_handler.setFormatter(formatter)\n",
    "\n",
    "# Create a logger and add the handlers\n",
    "logger = logging.getLogger(__name__)\n",
    "logger.addHandler(file_handler)\n",
    "# logger.addHandler(console_handler)\n",
    "\n",
    "try:\n",
    "    # Your code that may raise an exception goes here\n",
    "    1 / 0  # Example division by zero causing an exception\n",
    "except Exception as e:\n",
    "    # Log the exception\n",
    "    logger.error(f'Exception: {type(e).__name__} - Timestamp: {datetime.datetime.now()}')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99c2c69a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
